Convert paths, arguments, and env vars to Unicode for remote execution.

jmillikin opened the pull request:
Fixes #14381

I didn't touch the `Platform` construction because it seems to be used in parts of Bazel that aren't specific to remote execution. Maybe it'd be fine to change too? Comments welcomed.

coeuvre commented:
cc @tjgq

tjgq reviewed (changes requested):
It's not entirely obvious to me that `Platform` has the same issue (its contents originate from command line arguments, not Starlark). Adding a test would be the way to be sure, but feel free to punt on it, since non-ASCII values are probably not as common there.

FYI, while this PR is most definitely welcome, I suspect it doesn't _completely_ fix #14381, since that bug report also mentions an encoding issue in `bazel aquery`. So feel free to mention it, but I'd suggest leaving it open for now.

jmillikin commented:
> FYI, while this PR is most definitely welcome, I suspect it doesn't completely fix https://github.com/bazelbuild/bazel/issues/14381, since that bug report also mentions an encoding issue in `bazel aquery`. So feel free to mention it, but I'd suggest leaving it open for now.

I think the bad `bazel aquery` output might be the same bug, or one nearly identical. It's assembling a Unicode string for the output, and doesn't convert Starlark strings to Unicode before formatting them.

I've tweaked this PR so it also fixes `aquery`, using the new helper functions.

> It's not entirely obvious to me that `Platform` has the same issue (its contents originate from command line arguments, not Starlark).

While platform properties can be obtained from CLI flags, they also come from Starlark (the `platform` rule or `exec_properties` common attribute). I agree that non-ASCII values in platform properties are probably rare, so not going to worry about it for now, but it might be worth fixing in the future.

For reference, the following property configurations are equivalent:
```
test_unicode(
    name = "test_unicode",
    exec_properties = {"PROPERTY_あ": "あ"},
)

$ bazel build --remote_default_platform_properties='properties {name:"PROPERTY_あ" value:"あ"}'

$ bazel build --remote_default_exec_properties=PROPERTY_あ=あ
```

And result in the same (incorrect) `Platform` message being sent to the remote executor.
```
platform: {
  properties: {
    name: "PROPERTY_ã\u0081\u0082"
    value: "ã\u0081\u0082"
  }
}
```

jmillikin commented on src/test/shell/bazel/remote/remote_execution_test.sh:
```diff
@@ -3771,4 +3776,108 @@ EOF
+  inputs = glob(["inputs/**/*"]),
+)
+EOF
+  cat > rules.bzl <<EOF
```
Done.

jmillikin commented on src/test/shell/bazel/remote/remote_execution_test.sh:
```diff
@@ -3771,4 +3776,108 @@ EOF
+EOF
+  cat > rules.bzl <<EOF
+def _test_unicode(ctx):
+  out_dir = ctx.actions.declare_file(ctx.attr.name + "_outs/出力/あ.d")
```
Done -- great catch, thanks!

I've extended the test to verify the outputs have the expected structure + content.

jmillikin commented on src/test/shell/bazel/remote/remote_execution_test.sh:
```diff
@@ -3771,4 +3776,108 @@ EOF
+env | grep -v BASH_EXECUTION_STRING | grep TEST_UNICODE_ >> "\${report}"
+echo '' >> "\${report}"
+
+mkdir "\$1"
```
Done.

jmillikin commented on src/test/shell/bazel/remote/remote_execution_test.sh:
```diff
@@ -3771,4 +3776,108 @@ EOF
+)
+EOF
+
+  # inputs/入力/あ.txt
```
Done.

jmillikin commented on src/main/java/com/google/devtools/build/lib/remote/util/Utils.java:
```diff
@@ -707,4 +708,25 @@ public static void waitForBulkTransfer(
+   * Converts from a Starlark string, which contains raw bytes, to a Protobuf
+   * string, which contains Unicode code points.
+   */
+  public static String starlarkStringToProto(String starlarkString) {
```
Renamed to `starlarkToUnicode()` and `unicodeToStarlark()`.

I'm somewhat hopeful that a future iteration of Bazel might change Starlark strings from `[]byte` to `[]char`, so trying to avoid putting specific encodings into the function names.

jmillikin replied:
OK, I've renamed these to `decodeBytestringUtf8()` and `encodeBytestringUtf8()`. I also added some examples in the docstrings.

tjgq replied:
Thanks, this looks good now.

jmillikin commented on src/main/java/com/google/devtools/build/lib/remote/util/Utils.java:
```diff
@@ -707,4 +708,25 @@ public static void waitForBulkTransfer(
+  }
+
+  /** See {@link #starlarkStringToProto} */
+  public static String protoStringToStarlark(String protoString) {
```
Done.

jmillikin commented on src/main/java/com/google/devtools/build/lib/remote/merkletree/MerkleTree.java:
```diff
@@ -333,14 +334,17 @@ private static MerkleTree buildMerkleTree(
   private static FileNode buildProto(DirectoryTree.FileNode file) {
     return FileNode.newBuilder()
-        .setName(file.getPathSegment())
+        .setName(Utils.starlarkStringToProto(file.getPathSegment()))
```
Done.

jmillikin commented on src/test/shell/bazel/remote/remote_execution_test.sh:
```diff
@@ -3771,4 +3776,108 @@ EOF
 }
 
+function test_unicode() {
+  # Bazel can handle Unicode paths if it has either a UTF-8 or ISO-8859-1
```
I've expanded the comment -- is the new version clearer about which path encodings Bazel supports + locale requirements?

jmillikin replied:
Sounds good, I added that comment to the test and tweaked the other comments to be clearer about how Bazel uses the ISO-8859-1 locale for parsing UTF-8 filenames.

tjgq replied:
Much better, thanks.

jmillikin commented on src/test/shell/bazel/remote/remote_execution_test.sh:
```diff
@@ -3771,4 +3776,108 @@ EOF
+  # Bazel can handle Unicode paths if it has either a UTF-8 or ISO-8859-1
+  # locale available, but the in-tree remote execution worker has a hard
+  # requirement on UTF-8.
+  if ! "$is_windows"; then
```
On Windows, locales aren't needed to decode file paths. The requirement of a UTF-8 locale is only applicable to UNIX platforms with arbitrary bytes in paths (Linux, BSD).

tjgq replied:
I understand that Windows paths are Unicode, not arbitrary byte sequences. However, this test doesn't just deal in file *paths*, it also deals in file *contents* (namely, of .bzl files embedding non-ASCII characters).

Shouldn't we still test that Bazel will handle these files correctly *on Windows*? What happens if we remove `if ! $is_windows`? Is the Bash environment used by the integration tests complete enough to support `LC_ALL=en_US.UTF-8`?

jmillikin replied:
The Starlark spec requires UTF-8 encoding of `BUILD` and `.bzl` files:

https://github.com/bazelbuild/starlark/blob/master/spec.md#lexical-elements:
> A Starlark program consists of one or more modules. Each module is defined by a single UTF-8-encoded text file.

This has been a requirement since at least Bazel v0.17: https://docs.bazel.build/versions/0.17.1/skylark/spec.html#lexical-elements

I don't think Bazel enforces this, but from the discussion in https://github.com/bazelbuild/bazel/issues/4551 it sounds like a future version of Bazel might validate UTF-8 validity of inputs.

tjgq replied:
I agree with everything you wrote, but it doesn't answer my question. I'm asking whether it's possible to run this test on Windows as well (where AIUI the "store UTF-8 as Latin1 and convert it back on the way out" hack still applies, even assuming BUILD/bzl files are UTF-8 encoded per the spec).

It's fine if the answer is "it's impossible / too hard to make this test run on Windows due to platform limitations". I won't insist on it then.

jmillikin replied:
I don't understand your question. This test is running on Windows.

tjgq replied:
Nevermind, sorry. I misread the code.

tjgq reviewed (commented):
> I've tweaked this PR so it also fixes `aquery`, using the new helper functions.

Thanks! Do you mind also adding a test for `bazel aquery`, probably in src/test/shell/integration/aquery_test.sh?
 
> While platform properties can be obtained from CLI flags, they also come from Starlark (the `platform` rule or `exec_properties` common attribute). I agree that non-ASCII values in platform properties are probably rare, so not going to worry about it for now, but it might be worth fixing in the future.

Thanks, I wasn't aware of the `exec_properties` attribute. I agree that we don't have to fix this right now.

jmillikin commented:
> Thanks! Do you mind also adding a test for `bazel aquery`, probably in src/test/shell/integration/aquery_test.sh?

Done -- I added a test for `--output=text`.

Fixing the other `aquery` output formats (for example `--output=textproto`) are a bit beyond my depth in terms of Bazel knowledge. They seem to live mostly in `skyframe/actiongraph/` and I'm having some trouble figuring out what downstream consumers of their output might depend on exactly matching Bazel internal state.

For those, I've added a `DISABLED_` test that would pass if `--output=textproto` passed along Unicode. Hopefully it's OK to fix that in a separate change.

jmillikin commented:
Note on the `bazel aquery` output: on Windows, reliably writing non-ASCII to stdout requires a specially configured `PrintWriter`. I got lost trying to figure out how the `OutputStream` stream gets built, so instead it now escapes non-ASCII characters rather than try to print them. It's a little less pretty in the output, but is still usable for debugging and works on all tested platforms.

Also, apparently, on Windows strings for file paths are Unicode and strings from Starlark are UTF-8. I extended the logic in `decodeBytestringUtf8()` to try to handle this as best it can.

Here's how the new `aquery` output looks:
```
$ cat pkg/BUILD 
genrule(
    name = "bar",
    srcs = glob(["*.txt"]),
    outs = ["bar_out.txt"],
    cmd = "echo unused > $(OUTS)",
)
$ ls pkg
BUILD            ünïcödë.txt
$ bazel aquery --output=text //pkg:bar | grep Inputs
  Inputs: [external/bazel_tools/tools/genrule/genrule-setup.sh, pkg/{U+00FC}n{U+00EF}c{U+00F6}d{U+00EB}.txt]
```

tjgq commented on src/main/java/com/google/devtools/build/lib/query2/aquery/ActionGraphTextOutputFormatterCallback.java:
```diff
@@ -330,4 +333,34 @@ private Map<String, String> getParamFileNameToContentMap() {
+   * to be durable against unusual encoding settings, and does not guarantee
+   * that the escaping process is reverseable.
+   *
+   * Non-printable ASCII characters are formatted as `{U+00XX}`. Characters
```
Let's go with the standard-ish `\uXXXX` notation to make the output copy-pastable into a script.

jmillikin replied:
Done.

tjgq reviewed (approved):
Thanks, I'm going to import this now. Apologies for my confusion earlier and the review delay (I was out on Friday).
